druntime/src/core/internal/gc/impl/conservative/gc.d

```diff
@@ -1398,15 +1402,19 @@ class ConservativeGC : GC
         return __arrayStart(info)[0 .. usedSize];
     }
 
-    bool expandArrayUsed(void[] slice, size_t newUsed, bool atomic = false) nothrow
+    bool expandArrayUsed(void[] slice, size_t newUsed, bool atomic = false) nothrow @trusted
```

Comment:
silly question: why is this whole function `@trusted`?

Reply:
I suppose it doesn't have to be, but the parts that need trusted are most of the function. We are dealing with block information, pointers, slice.ptr, sizes. Very little would be not trusted. I think it would be worse to have trusted lambdas everywhere.

Reply:
Trusted lambdas would be worse, yes, but that's far from the only available option. Private helper functions to encapsulate the unsafe code or even just documenting why the `@system` code is actually `@safe` are both far better options, imo.

I'm leaning towards the former option in this case, since the unsafe bits seem to be repeated almost verbatim a few times throughout this PR. It would significantly reduce the surface area of the unsafe code and dramatically improve reviewability and maintainability as a result.

Reply:
The biggest trusted thing is getting a slice ptr. Getting the block info for a pointer can be safe, whether it's from the cache or from the GC.

I'll see if I can rework the functions.

Reply:
Given that we are dealing with `BlkInfo`, which is inherently an unsafe structure (you have writeable access to the pointer and length), I think it's better to put this whole thing in `@trusted`. I'll see if I can abstract out some of this code, but I'm not sure it makes things better.

Reply:
OK, I'm pretty convinced the only way to make this more safe and less trusted is to encapsulate the BlkInfo into a more restrictive structure. And I'm not going to do that in this PR.

Reply:
OK, if you could add the summary of this review as a comment that would be great, but not absolutely necessary if you plan to rectify it later.